Multi layer aanpassingen

raymondedwards95 opened the pull request:
Ik heb de code wat opgeruimd en verbeterd. 
Nu geeft de iris-dataset steeds minder dan 5 fout.
mnist-data moet nog geprobeerd worden (geeft nog vaak overflow-errors).

raymondedwards95 committed:
aanpassingen docstrings

raymondedwards95 committed:
opmaak

raymondedwards95 committed:
functies voor aanpassen van train-data

raymondedwards95 committed:
aanpassingen aan predict()

Vooral verduidelijking

raymondedwards95 committed:
bepaal_succes() toegevoegd

functie om het aantal foute voorspellingen te geven

raymondedwards95 committed:
aanpassing predict() - schaling

raymondedwards95 committed:
aanpassing functie train()

mooier gemaakt

raymondedwards95 committed:
Kleine aanpassingen aan bestand

raymondedwards95 committed:
weghalen van niet-nuttige regels

raymondedwards95 committed:
index in een loop verandert

raymondedwards95 committed:
opschonen code

normalisaties aangepast: deze worden nu pas toegepast in de functies
voor het bepalen van uitvoer
Daarnaast ook self.w en self.b weggehaald: deze zitten nu in self.p
(eerste helft is w, andere helft is b)
Met deze laatste update, is de fout bij irisdata nog maar een paar
procent (<5 fout).
MNIST data moet nog getest worden

raymondedwards95 committed:
verandering aan w en b

standaard bias is nu 0
standaard gewicht zit tussen 0 en factor
een lagere factor helpt bij mnist

MarikenWeijs reviewed (commented):
Hee Raymond, er is echt vet veel gebeurd, super goed. Het ziet er overzichtelijk uit. De manier van opslaan voor import en export snap ik nog steeds niet, maar de rest lijkt me prima. Ik denk dat we hem met de master kunnen samenvoegen.

raymondedwards95 committed:
maximale waarde voor gewichten

Door de gewichten (en bias) kleiner te maken, lukt het om de mnist data
werkend te krijgen (als in: er gebeurt iets)

raymondedwards95 commented:
Voor import en export is het vooral documentatie lezen (https://docs.scipy.org/doc/numpy/reference/routines.io.html). 

Verder lukt het een beetje met de mnist data (zie laatste (_nieuwe_) commits). Door de initiële parameters (gewichten, bias) kleiner te maken (door middel van een factor, die momenteel als argument ingevoerd moet worden), lukt het om een netwerk met mnist te trainen. 
Het grootste probleem is dat het lang duurt om alle (_60000 voorbeelden_) data langs te gaan. Daarnaast is het lastig om goede parameters (aantal neuronen, aantal lagen, aantal iteraties, maximale waarde gewichten en alfa) te vinden.

raymondedwards95 committed:
factor toegevoegd

MarikenWeijs committed:
niewe_testdata

aangepast zodat er nieuwe data aan de oude wordt toegevoegd ipv erbij
wordt opgeteld

raymondedwards95 committed:
lezen van IrisData in een definitie gezet

verder nog wat kleine aanpassingen

raymondedwards95 committed:
Een leeg NeuralNetwork geeft nu standaard een 'and'-poort

raymondedwards95 committed:
bepaling goede parameters (?)

de vooraf gedefinieerde 'factor' en 'alfa' werken bij dit specifieke
geval (en een aantal andere). Er zit niet echt iets achter deze waarden,
behalve dat er resultaten zijn te verkrijgen (25% fout bij het oefenen
met de helft (30000 voorbeelden) van de testdata).

raymondedwards95 commented:
Ik heb nu de mnist data getest en een netwerk ermee getraind. Nu krijg ik na 25000 iteraties (duurde zo lang...) ongeveer 26% fout.

gebruikte waarden: f=0.0357, alfa=1.19e-6, k=[9,5]
eind gewichten en biassen zijn opgeslagen, als je ze wil hebben, kan ik ze sturen of hier op github neer zetten.

Ik kreeg wel een iets beter resultaat (25% fout) wanneer ik maar met de helft van de testdata ging trainen.

MarikenWeijs commented:
zou chill zijn als je je resultaten hier neer zou kunnen zetten Raymond

raymondedwards95 commented:
# Resultaten
### Samenvatting:
resultaten met een verschillend aantal testdata:
Boven: 30000 en 5000
Onder: 60000 en 10000

Overige gegevens: 
- k = [9,5]
- f = 1 / (len(k) * sum(k)) (_vraag niet waarom dit werkt, ik weet ook niet zo goed of het met meer elementen in k ook nog werkt_)
- alfa = len(k) / (aantal * sum(k)) (_ook hier weet ik niet precies waarom het goed gaat_)

Het zijn geen optimale parameters, maar in dit geval doen ze wat er moet gebeuren (trainen van het netwerk).

![mnist-resultaat](https://cloud.githubusercontent.com/assets/26598884/25223067/a10575ca-25ba-11e7-9ba1-82d14565f55b.PNG)


### Parameters:
[mnist-parameters.zip](https://github.com/hannahgathu/PIDW-Eindopdracht/files/941881/mnist-parameters.zip)

Om de parameters te gebruiken, moet je ze uit het zip-bestand halen, en dan importeren in een leeg neuraal netwerk met import_parameters(bestand).
Dus:
`n = MultiLayerNeuralNetwork()`
`n.import_parameters('NN-mnist-60000.npz')`
`n.bepaal_succes(n.X, n.Y)`

# Code
### Import/Export
Het idee is dat je alle gegevens (_gewichten, bias, normalisaties..._) in een extern bestand stopt, om ze later terug te kunnen halen (zie voorbeeld hierboven). _Numpy_ heeft een systeem om getallen en arrays in een binair bestand te plaatsen (`numpy.save` of afgeleiden daarvan). Bij het opslaan koppelt Numpy de elementen aan een bepaalde key, zoals in een directionary. Daarom staan er steeds argumenten als `l=self.l` in save. Hier wordt de sleutel 'l' gekoppeld aan waarde `self.l`.
Wil je de data importeren, dan moet het bestand worden geladen (`data = numpy.load(file)`). En dan gebruik je de sleutels om de getallen en arrays eruit te halen (`self.l = data['l'])


### Oude code
Is het al mogelijk om oude code (_single layer, two layer..._) te archiveren?

edit: bestand moet een bestandsnaam/string zijn

raymondedwards95 committed:
def main() weggehaald

Nu kan import MultiLayerNeuralNetwork veilig worden gebruikt (dus `from
"file" import "class"` is niet perse nodig)

raymondedwards95 committed:
kleine aanpassing aan import/export

iets met bestandsnamen

raymondedwards95 committed:
kleine fix bij import/export

MarikenWeijs committed:
comment weg

MarikenWeijs committed:
comments weg die overbodig zijn geworden

MarikenWeijs committed:
bereken_y gebruikt in predict

ipv code kopiëren

MarikenWeijs committed:
docstring in derde persoon enkelvoud
